packages/aws-cdk-lib/aws-stepfunctions/lib/state-machine.ts

```diff
@@ -452,6 +476,50 @@ export class StateMachine extends StateMachineBase {
       }
     }
 
+    if (props.kmsKey) {
+      this.role.addToPrincipalPolicy(new iam.PolicyStatement({
+        effect: iam.Effect.ALLOW,
+        actions: [
+          'kms:Decrypt', 'kms:GenerateDataKey',
+        ],
+        resources: [`${props.kmsKey.keyArn}`],
+        conditions: {
+          StringEquals: {
+            'kms:EncryptionContext:aws:states:stateMachineArn': Stack.of(this).formatArn({
+              service: 'states',
+              resource: 'stateMachine',
+              sep: ':',
+              resourceName: this.physicalName,
+            }),
+          },
+        },
+      }));
+
+      if (props.logs && props.logs.level !== LogLevel.OFF) {
+        this.role.addToPrincipalPolicy(new iam.PolicyStatement({
+          effect: iam.Effect.ALLOW,
+          actions: [
+            'kms:Decrypt', 'kms:GenerateDataKey',
+          ],
+          resources: [`${props.kmsKey.keyArn}`],
+          conditions: {
+            StringEquals: {
+              'kms:EncryptionContext:SourceArn': Stack.of(this).formatArn({
+                service: 'logs',
+                resource: '*',
```

Comment:
I don't like using `*` but after speaking with KMS project lead it looks like CWL doesn't currently support matching the log group resource in encryption context for `EncryptionContext:SourceArn`.